### ISSUE TITLE
Error in documentation on macro hygiene

### ISSUE THREAD

**Issue by jlapeyre** [JuliaLang/julia#45323]
The documentation on macro hygiene says explicitly that with this macro
```julia
macro mytime1(ex)
    return quote
        local t0 = time_ns()
        local val = $ex
        local t1 = time_ns()
        println("elapsed time: ", (t1-t0)/1e9, " seconds")
        val
    end
end
```
the code `@mytime1 t0 = 0` will work as intended.
But, it doesn't:
```julia
julia> @mytime1 (t0 = 0);
elapsed time: 4.576732909746654e6 seconds
```
However, writing  `local val = $(esc(ex))` *does* make the example work as intended:
```julia
macro mytime2(ex)
    return quote
        local t0 = time_ns()
        local val = $(esc(ex))
        local t1 = time_ns()
        println("elapsed time: ", (t1-t0)/1e9, " seconds")
        val
    end
end

julia> @mytime2 (t0 = 0);
elapsed time: 7.52e-7 seconds
```

The documentation goes on to say that doing `$(esc(ex))`, as in `@mytime2`, is indeed necessary for a robust macro; but for other reasons. To emphasize again, it explicitly says that " assignment[s] to a variable called `t0`" does not require `esc`.

**Comment by vtjnash** [comment]
I agree. It is confusing to first show the incorrect snippet, then later explain how to write correctly

**Comment by jlapeyre** [comment]
> It is confusing to first show the incorrect snippet, then later explain how to write correctly

If I understand your comment correctly, it's worse than that. It essentially says:

Here is a macro with no `esc` that introduces a variable `t0` in the body. But what if the expression  passed by the caller also assigns to a variable called `t0`?  How will they be made distinct? No worry; the rules when to gensym a variable already prevent this from happening, *without* needing to use `esc`.

But, in fact, the `esc` *is* necessary in this particular case.

**Comment by KristofferC** [comment]
> But, in fact, the esc is necessary in this particular case.

I don't think so, esc is not related to that. You need the esc to not apply hygiene to the input expression (so the name of the macro expanded variable is the same as the input)

**Comment by jlapeyre** [comment]
I keep rereading it, trying to interpret it as unclear rather than wrong. But, it is fairly clear

1. Introduce macro with user input expression *not* escaped
2. State that there are two potential problems prob 1) user input `ex` assigns to a "variable called `t0`". prob 2) "defined its own time_ns variable."
3. State that the macro expander solves the problem in the following way:
4. Give the rules for when to apply hygiene to parts of the expression. I interpret this as the entire "way" referred to in 3. There is a "First" and a "...then". So two parts. You might want to say that the text several lines below about `esc` are also part of the "way". But the next item, number 5, makes it clear that there is no more to the "way", we are done. (of course later we find that the "way" is not really enough. But, is enough for the two problems at hand.)
5. Declare that this solves the two problems: "both of the above concerns are handled".  prob 1) " macro's locals will not conflict with any user variables" and prob 2) "time_ns and println will refer to the Julia Base definitions" [I added the numbers here as the my interpretation of what refers to what]
6. State that "one" problem remains. You might think that this means either one of the two problems remain, or that there is a third heretofore unconsidered problem. And "one problem remains" seems to imply the first meaning, not the second. Maybe. But, based on all the text above, I interpret this to mean we consider now a third problem. 
7.  Give an example of the "remaining" problem. Then, for the first time, introduce `esc`

**Comment by jlapeyre** [comment]
I'm not arguing that the rules stated are wrong. I think the following agrees with the rules for when to rename variables. Note that the following is with *no* escape. Because the discussion in the doc is premised on no `esc` yet.
```julia
julia> striplines(macroexpand(Main, :(@mytime1 t0=1)))
quote
    local var"#139#t0" = Main.time_ns()
    local var"#138#val" = (var"#139#t0" = 1)
    local var"#140#t1" = Main.time_ns()
    Main.println("elapsed time: ", (var"#140#t1" - var"#139#t0") / 1.0e9, " seconds")
    var"#138#val"
end

julia> striplines(macroexpand(Main, :(@mytime1 global t0=1)))
quote
    local Main.t0 = Main.time_ns()
    local var"#141#val" = global t0 = 1
    local var"#142#t1" = Main.time_ns()
    Main.println("elapsed time: ", (var"#142#t1" - Main.t0) / 1.0e9, " seconds")
    var"#141#val"
end
```
But, they don't prevent the "user" variables and the macros variables from colliding. It's even kinda clear for why both `t0`s are renamed to the same thing. The doc refers to the "variables within a macro result", which I interpret to mean regardless of where they came from.

I suppose the entire problem and solution up to this point has been about protecting the variables defined by the macro. And I agree, they have been solved.

Then the third thing, solved by `esc` is preventing the things you don't want from being renamed. If this is what the doc means, then yes, I suppose it's unclear rather than wrong. I mean, the text up to the point of "one problem remains" should make it clear that if I pass a `t0` in my expression it is going to collide with `t0` in the macro.

**Comment by vtjnash** [comment]
I agree with you
